schema/shared/src/main/scala/zio/blocks/schema/DynamicValue.scala

```diff
@@ -173,4 +173,131 @@ object DynamicValue {
+        if (old == n) DynamicPatch.empty
+        else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.LongDelta(n - old))))
+
+      case (PrimitiveValue.Double(old), PrimitiveValue.Double(n)) =>
+        if (old == n) DynamicPatch.empty
+        else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.DoubleDelta(n - old))))
+
+      case (PrimitiveValue.Float(old), PrimitiveValue.Float(n)) =>
+        if (old == n) DynamicPatch.empty
+        else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.FloatDelta(n - old))))
+
+      case (PrimitiveValue.Short(old), PrimitiveValue.Short(n)) =>
+        if (old == n) DynamicPatch.empty
+        else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.ShortDelta((n - old).toShort))))
+
+      case (PrimitiveValue.Byte(old), PrimitiveValue.Byte(n)) =>
+        if (old == n) DynamicPatch.empty
+        else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.ByteDelta((n - old).toByte))))
+
+      case (PrimitiveValue.BigInt(old), PrimitiveValue.BigInt(n)) =>
+        if (old == n) DynamicPatch.empty
+        else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.BigIntDelta(n - old))))
+
+      case (PrimitiveValue.BigDecimal(old), PrimitiveValue.BigDecimal(n)) =>
+        if (old == n) DynamicPatch.empty
+        else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.BigDecimalDelta(n - old))))
+
+      // String diff
+      case (PrimitiveValue.String(old), PrimitiveValue.String(n)) =>
+        if (old == n) DynamicPatch.empty
+        else {
+          val ops = StringOp.diff(old, n)
+          if (ops.isEmpty) DynamicPatch.empty
+          else DynamicPatch.single(DynamicPatchOp(optic, Operation.PrimitiveDelta(PrimitiveOp.StringEdit(ops))))
+        }
+
+      // Different primitive types or non-diffable - replace
+      case _ =>
+        DynamicPatch.single(DynamicPatchOp(optic, Operation.Set(Primitive(newPrim))))
+    }
+  }
+
+  private def diffRecords(
+    optic: DynamicOptic,
+    oldFields: Vector[(String, DynamicValue)],
+    newFields: Vector[(String, DynamicValue)]
+  ): DynamicPatch = {
+    var ops = Vector.empty[DynamicPatchOp]
+    val oldMap = oldFields.toMap
+    val newMap = newFields.toMap
+    val allKeys = (oldMap.keySet ++ newMap.keySet).toVector
+
+    for (key <- allKeys) {
+      val fieldOptic = optic.field(key)
+      (oldMap.get(key), newMap.get(key)) match {
+        case (Some(oldVal), Some(newVal)) =>
+          val fieldPatch = diffInternal(fieldOptic, oldVal, newVal)
+          ops = ops ++ fieldPatch.ops
+        case (None, Some(newVal)) =>
+          ops = ops :+ DynamicPatchOp(fieldOptic, Operation.Set(newVal))
+        case (Some(_), None) =>
+          // Field removed - this is structural, use Set with empty record marker
```

Comment:
This actually argues that you need a DeleteField operation. Otherwise this will not be lawful in cases where the right hand side loses fields.

Now, for typed schemas, maybe this is used and maybe it isn't, but DynamicPatch should be well-defined by itself without reference to a typed later atop.